### PR TITLE
drivers: add hts221 support

### PR DESCRIPTION
**Pull request by hoihu**
This is my attempt of an "API compatible" sensor driver which may serve as a discussion base for #3218. It's a dual sensor device that measures humidity and temperature and can be configured in various ways. It has even an in-built heater. I believe it serves as a good test and discussion base for the new API. The datasheet is located here http://www.st.com/en/mems-and-sensors/hts221.html. 

The HTS221 sensor is present on the SenseHat board for the raspberry Pi (which can be easily reused to work with the pyboard).

The driver supports fixed and floating point targets, with the conversion functions that was proposed in #2093. I specifically paid attention to the precision of the output values (e.g. so that temperature are not returned with 10 digits precision). 

The buffer is preallocated in the constructor, so that during runtime there should be no memory allocation necessary.

I have not completly followed the guidelines in #2093, for the following reasons:
- the `fixed`statement that was proposed to be present for every `get_xxx`accessor method is replaced by an equivalent argument for the constructor of the driver (hence the method definiton is slightly shorter). This makes sense, since the user hardly wants to use both fixed and floating point support of the same driver for a given Platform
- the `get_temperature`and `get_humidity`accessor functions are included (not inherited from a `Sensor Base` class) since this allows introspection at the REPL and in general makes the code more readable IMO
-  I omitted the generic `config` method and just used e.g. a `set_heater`function - for the same reasons given above
 
The reasoning above is of course subject for discussing and just my point of view. Any feedback is very much welcome. I will upgrade my other drivers for the SenseHat once the Sensor API is considered stable.

### CODE REVIEW

File: drivers/i2c/hts221.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        self.t_slope = (1000 * (self.t1_deg - self.t0_deg)) // (self.calib.t1_out - self.calib.t0_out)
+        self.h_slope = (10000 * (self.calib.h1_rh - self.calib.h0_rh)) // (self.calib.h1_out - self.calib.h0_out)
+        
+    def measure(self):
+        # enable one-shot measurement
+        self.i2c.writeto_mem(self.addr, 0x21, b'\x01')
+        self.measure_done = False
+
+    def measure_end(self):
+        # blocking, timeout after 2sec
```
Comment (peterhinch) [inline drivers/i2c/hts221.py]: Is this comment correct?

**Comment by dpgeorge** [comment]
Thanks @hoihu for for pushing forward with this discussion on sensors.  It is a neat idea to have fixed vs float configurable in the constructor and the get_XXX() methods return one or the other.  If a user wanted to use both fixed and float then they would select fixed results and convert them to float when needed using fixed_to_float().

File: drivers/i2c/hts221.py (hunk 1 above)
Comment (hoihu) [inline drivers/i2c/hts221.py]: thanks for pointing out - yes it's not correct. It should be 0.2sec.

Actually the datasheet is a bit unclear about the sampling time in one-shot mode, but I've measured it to be always smaller than 10msec.

**Comment by hoihu** [comment]
I've pushed some small changes (eliminating the unneeded `array` import and some documentation cleanup. 

Some thoughts:
- location of the drivers: Maybe a separate `micropython-drivers` repository? Or part of the main uPy repository within `/drivers`? I'd tend towards a separate repository.
- upip installation support? 
- docs: How do we handle things like links to data sheets, example REPL sessions, maybe a simple wiring diagram? If the repository is separate, maybe it makes sense to setup a RTD sphinx doc and each driver can freely describe itself within an .rst file. This may be possbile too within the existing sphinx documentation. The code can then be freed from these kind of documentations, making the driver's memory footprint smaller.

**Comment by pfalcon** [comment]
> It is a neat idea to have fixed vs float configurable in the constructor and the get_XXX() methods return one or the other.

And I don't think it's neat. It's just another instance variable which is not related to the device functionality per se. A particular application will want either fixed or float, so can call get_fixed() or get_float() explicitly. Whereas here it's a constant stored as a variable, taking memory from user's app.

**Comment by turbinenreiter** [comment]
Spell it out, we need to know how it would look in real code.

`accelerometer.get_acceleration()`
`barometer.get_pressure()`

Vs.

`accelerometer.get_float(accelerometer.ACCELERATION)`
`barometer.get_fixed(barometer.PRESSURE)`

**Comment by hoihu** [comment]
Some thoughts on sensor measurment intervals and blocking / non-blocking mode:

Some sensors have the ability to measure in a pre-configured interval (e.g. 10Hz) without an external trigger. Examples of such sensors include temperature/humidty sensors but also (more importantly) gyro sensors. 

For gyro sensors, it's important to have a fixed, constant measurement rate since the follow-up fusion algorithms assumes this. Often, gyro sensors also have buffers to store sensor data locally, e.g. in a FIFO.

For theses kind of sensors, `measure`can immediately return: The overhead of issuing an "I2C one-shot command" is avoided. For low power application this seems like a sensible solution.

The only disadvantage I can think of is that you can't predict the blocking time in `measure_end`since that depends on the configured (external) measurement rate and the time of the last measurement. A "one-shot" setup would be more deterministic.

Perhaps yielding is anyway not a bad solution for the blocking `while`loop in `measure_end`. The uasyncio scheduler may benefit from that (but I have to admit I'm no expert on that subject)

And regarding FIFO. It would be nice to make these FIFO values accessible via generator methods so a user can just iterate over them. Maybe we could name them `iter_xxx`:
```
for value in sensor.iter_pressure():
    print("{}".format(value))
```

**Comment by hoihu** [comment]
> It's just another instance variable which is not related to the device functionality per se

> ...or can call get_fixed() or get_float() explicitly

But I guess the downside is readabilty of the code, as @turbinenreiter pointed out. Or can you further elaborate on that, perhaps with example code?

**Comment by turbinenreiter** [comment]
`measure` could return the remaining time until done, but I'm not sure if that works for all cases.

**Comment by hoihu** [comment]
> measure could return the remaining time until done, but I'm not sure if that works for all cases

you can't do that if the sensor initiates the measurements by itself. Even if you know the sampling rate - you don't know when it has started the last measurement.

**Comment by deshipu** [comment]
I wrote a whole bunch of sensor drivers for MicroPython not waiting for an "official" API, and the way I got around the float problem was by introducing an optional `raw` parameter for the `get_*` methods. If set to true, the method simply returns the value it got from the sensor directly, without any scaling or unit conversions, letting the user handle it how they need it. This is useful not only when floats are not available on the given port, but also with some advanced use or when you don't want to lose any accuracy. And it's super-helpful for debugging.

One of the drivers using that: https://github.com/adafruit/micropython-adafruit-tcs34725/blob/master/tcs34725.py#L100

**Comment by dpgeorge** [comment]
A different hts221 driver was merged in fc6d989f8cca61b33715fcf2e8f1eaa0d0ed67ec, which is a bit simpler than this PR.

Unfortunately there is still no standard driver/sensor API.

**Pull request closed without merging**